flink-runtime/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceAsyncImpl.java

```diff
@@ -125,13 +137,14 @@ protected void foreachTimer(
                 "Batch operation is not supported when using async state.");
     }
 
-    private void maintainContextAndProcess(
+    private StateFuture<Void> maintainContextAndProcess(
             InternalTimer<K, N> timer, ThrowingRunnable<Exception> runnable) {
         RecordContext<K> recordCtx = asyncExecutionController.buildContext(null, timer.getKey());
```

Comment:
Apart from current topic: How about make timer as the record and invoke `buildContext`?